### PR TITLE
plugins/sto: Revamp plugin

### PR DESCRIPTION
**Pull request by davidebeatrici**
This commit changes the code so that it follows the style of the other plugins.

The offsets are not updated, this means that the plugin only works with a very old version of the game, but with the new code it should be much more easy for a contributor to update them.

The commit also gets rid of a strict-aliasing rules break warning, which appeared when compiling with MinGW, caused by:

`if (* reinterpret_cast<DWORD *>(& posblock[52]) == 1) {`

because we're reading a spawn state or magical value, which is an unsigned char, from an array of floats.

Fixes issue #2757.

### CODE REVIEW

**Comment by Kissaki** [comment]
> The offsets are not updated, we had to do that

Do what?

**Comment by Kissaki** [comment]
Please make the commit describe what this logically does.

The line that threw a warning is

> if (* reinterpret_cast<DWORD *>(& posblock[52]) == 1) {

where `posblock[52]` seems to be spawn state.

So the issue here is that we read data into a float array, but this is not a float.

That should be a pretty simple change.

And that change should be in a separate commit to changing the version string for example.

**Comment by davidebeatrici** [comment]
> Do what?

Revamp the plugin, I'm going to specify it in the commit message.

> Please make the commit describe what this logically does.
The line that threw a warning is
if (* reinterpret_cast<DWORD *>(& posblock[52]) == 1) {
where posblock[52] seems to be spawn state.
So the issue here is that we read data into a float array, but this is not a float.
That should be a pretty simple change.
And that change should be in a separate commit to changing the version string for example.

Indeed, but I think that it's better if we entirely change the plugin so that it follows the style of other ones.

File: plugins/sto/sto.cpp [hunk 1]
```diff
@@ -23,71 +28,56 @@ static void inline u8(std::wstring &dst, const std::string &src) {
...
+
+	// Boolean value to check if game addresses retrieval is successful
+	bool ok;
+	// Char values for extra features
+	char version[17], identity_block[0x200], context_block[0x80];
+	// State
+	unsigned char state;
+
+	// Peekproc and assign game addresses to our containers, so we can retrieve positional data
+	ok = peekProc(pModule + state_offset, &state, 1)							&& // Magical state value: 1 when spawned.
```
Comment (mkrautz) [inline plugins/sto/sto.cpp]: In the original version, this was a DWORD (i.e., uint32_t). I'd include <stdint.h> and use uint32_t instead.

File: plugins/sto/sto.cpp [hunk 2]
```diff
@@ -23,71 +28,56 @@ static void inline u8(std::wstring &dst, const std::string &src) {
...
+			peekProc(pModule + position_offset, avatar_pos, 12)					&& // Avatar Position values (X, Y and Z).
+			peekProc(pModule + position_offset + 0xC, camera_pos, 12)			&& // Camera Position values (X, Y and Z).
+			peekProc(pModule + position_offset + 0x3C, avatar_front, 12)		&& // Front vector values (X, Y and Z).
+			peekProc(pModule + position_offset + 0x48, camera_front, 12)		&& // Top vector values (X, Y and Z).
```
Comment (mkrautz) [inline plugins/sto/sto.cpp]: Is this comment correct? Reads into camera_front, but says "top vector values"?

File: plugins/sto/sto.cpp (hunk 1 above)
Comment (mkrautz) [inline plugins/sto/sto.cpp]: Please fix the tabs/indentation of the `&&`s. It looks very odd.

File: plugins/sto/sto.cpp (hunk 2 above)
Comment (davidebeatrici) [inline plugins/sto/sto.cpp]: I forgot to edit it when I copied the part from another plugin.

File: plugins/sto/sto.cpp (hunk 1 above)
Comment (davidebeatrici) [inline plugins/sto/sto.cpp]: How should I indent it?

Comment (davidebeatrici) [inline plugins/sto/sto.cpp]: I don't think that the value from the game could be bigger than an unsigned char (255), but I'm going to change it anyway, as the original plugin used DWORD.

**Comment by davidebeatrici** [comment]
We decided to make STO a NULL plugin, as we can't know if our changes are right.

**Comment by davidebeatrici** [comment]
PR: https://github.com/mumble-voip/mumble/pull/2941

**Pull request closed without merging**